pkg/server/tls_windows.go

```diff
@@ -0,0 +1,342 @@
+		return err
+	}
+	cc, err := clientIdentity.Certificate()
+	if err != nil {
+		return err
+	}
+	c.serverCert = sc
+	c.serverSigner = signer
+	c.clientRootCA = cc
+	c.serverIdentity = serverIdentity
+	return nil
+}
+
+func (c *winCertStoreHandler) findServerCertificate() (certstore.Identity, error) {
+	return c.findCertificate(c.cfg.ServerSystemStore, c.cfg.ServerStore, c.cfg.ServerIssuerCommonNames, c.cfg.ServerTemplateID, nil, c.getStore)
+}
+func (c *winCertStoreHandler) findClientCertificate() (certstore.Identity, error) {
+	return c.findCertificate(c.cfg.ClientSystemStore, c.cfg.ClientStore, c.cfg.ClientIssuerCommonNames, c.cfg.ClientTemplateID, c.subjectRegEx, c.getStore)
+}
+
+func (c *winCertStoreHandler) getStore(systemStore string, storeName string) (certstore.Store, error) {
+	st, err := certstore.StringToStoreType(systemStore)
+	if err != nil {
+		return nil, err
+	}
+	store, err := certstore.OpenSpecificStore(st, storeName)
+	if err != nil {
+		return nil, err
+	}
+	return store, nil
+}
+
+type getStoreFunc func(systemStore, storeName string) (certstore.Store, error)
+
+func (c *winCertStoreHandler) findCertificate(systemStore string, storeName string, commonNames []string, templateID string, subjectRegEx *regexp.Regexp, getStore getStoreFunc) (certstore.Identity, error) {
+	var store certstore.Store
+	var validIdentity certstore.Identity
+	var identities []certstore.Identity
+	// Lots of cleanup here that point to windows handles.
+	defer func() {
+		if store != nil {
+			store.Close()
+		}
+		// Now lets clean up any identities that are NOT the one we want
+		for _, i := range identities {
+			if i != validIdentity {
+				i.Close()
+			}
+		}
+	}()
+	store, err := getStore(systemStore, storeName)
+	if err != nil {
+		return nil, err
+	}
+	identities, err = store.Identities()
+	filtered, err := c.filterByIssuerCommonNames(identities, commonNames)
+	if err != nil {
+		return nil, err
+	}
+	filtered, err = c.filterByTemplateID(filtered, templateID)
+	if err != nil {
+		return nil, err
+	}
+	filtered, err = c.filterBySubjectRegularExpression(filtered, subjectRegEx)
+	if err != nil {
+		return nil, err
+	}
+	if len(filtered) == 0 {
+		return nil, fmt.Errorf("no certificates found")
+	}
+	// order oldest to newest
+	sort.Slice(filtered, func(i, j int) bool {
+		// Already accessed this so the error will not happen
+		a, _ := filtered[i].Certificate()
+		b, _ := filtered[j].Certificate()
+
+		return a.NotBefore.Before(b.NotBefore)
+	})
+
+	// Grab the most recent valid one
+	for i := len(filtered) - 1; i >= 0; i-- {
+		fc, _ := filtered[i].Certificate()
+		if time.Now().Before(fc.NotAfter) && time.Now().After(fc.NotBefore) {
+			validIdentity = filtered[i]
+			break
+		}
+	}
+	if validIdentity == nil {
+		return nil, fmt.Errorf("no certificates found")
+	}
+
+	return validIdentity, nil
+}
+
+func (c *winCertStoreHandler) filterByIssuerCommonNames(input []certstore.Identity, commonNames []string) ([]certstore.Identity, error) {
```

Comment:
The filter functions are kind of repetitive, can we make a generic filter function that these can use?

```go
func (c *winCertStoreHandler) filterCerts(in []certstore.Identity, f func(*x509.Certificate) bool} ([]certstore.Identity, error) {
  res := make([]certstore.Identity, 0, len(in)) 
  for _, ident := range in {
    cert, err := ident.Certificate()
    if err != nil {
      return err 
    }
    if f(cert) {
      res = append(res, ident)
    }
  }
  return res
}
```

Reply:
I originally played around with something similar, but I think it reduced clarity. For instance, you need to pass in two functions (or add the check in the caller) to see if you need to do any matching. I could move them out of inline funcs but the gain is minimal.

Will think on if there is a cleaner path. 

```
c.filterCertificate(identities, func() bool{
	return len(commonNames) != 0
}, func(cert *x509.Certificate) bool {
	for _, cfgName := range commonNames {
		if cert.Issuer.CommonName == cfgName {
			return true
		}
	}
	return false
})


func (c *winCertStoreHandler) filterCertificate(input []certstore.Identity, shouldMatch func() bool, match func(certificate *x509.Certificate) bool) ([]certstore.Identity, error) {
	if !shouldMatch() {
		return input, nil
	}
	returnIdentities := make([]certstore.Identity, 0)
	for _, identity := range input {
		cert, err := identity.Certificate()
		if err != nil {
			return nil, err
		}
		if match(cert) {
			returnIdentities = append(returnIdentities, identity)
		}
	}
	return returnIdentities, nil
}
```